scylla/src/transport/topology.rs

```diff
@@ -870,6 +871,106 @@ async fn query_user_defined_types(
+            PreCqlType::Tuple(types) => types
+                .iter()
+                .for_each(|type_| do_with_referenced_udts(what, type_)),
+            PreCqlType::UserDefinedType { name, .. } => what(name),
+        }
+    }
+
+    let defs = mem::take(udts);
+
+    // Build an indegree map: for each node in the graph, how many directly depending types it has.
+    let mut indegree = defs
+        .into_iter()
+        .map(|def| {
+            (
+                (def.keyspace_name.clone(), def.type_name.clone()),
+                (def, Cell::new(0u32)),
+            )
+        })
+        .collect::<HashMap<_, _>>();
+
+    // For each node in the graph...
+    for (def, _) in indegree.values() {
+        let mut increment_referred_udts = |type_name: &str| {
+            let deg_cell = indegree
+                .get(&(def.keyspace_name.clone(), type_name.to_string()))
+                .map(|(_, count)| count)
+                .unwrap();
+            deg_cell.set(deg_cell.get() + 1);
+        };
+        // For each type referred by the node...
+        for field_type in def.field_types.iter() {
+            let pre_cql_type = map_string_to_cql_type(field_type);
+            // If the type referred is another node in the graph...
+            if let Ok(type_) = pre_cql_type {
+                do_with_referenced_udts(&mut increment_referred_udts, &type_);
+            }
+        }
+    }
+
+    let mut leaves_first = vec![];
+    // The "dual" hash map serves as a temporary storage in order to overcome the limitation of drain_filter() being unstable.
+    let mut indegree_dual = HashMap::with_capacity(indegree.len());
+
+    // While there are any nodes remaining in the graph...
+    while !indegree.is_empty() {
+        let extend_beg = leaves_first.len();
+
+        // Remove nodes with indegree zero (i.e. leaves)
+        for (key, (def, deg)) in indegree.drain() {
+            if deg.get() == 0 {
+                leaves_first.push(def);
+            } else {
+                indegree_dual.insert(key, (def, deg));
+            }
+        }
+
+        if leaves_first.len() == extend_beg {
+            panic!(
+                "Bug in fetched User Defined Types definitions: circular type dependency detected.
+                 Topological sort is thus impossible. Panicking to prevent an infinite loop."
+            )
+        }
+
+        mem::swap(&mut indegree, &mut indegree_dual);
+
+        for new_leaf in &leaves_first[extend_beg..] {
+            let mut decrement_referred_udts = |type_name: &str| {
+                *indegree
+                    .get_mut(&(new_leaf.keyspace_name.clone(), type_name.to_string()))
+                    .unwrap()
+                    .1
+                    .get_mut() -= 1;
+            };
+            // For each node that the currently removed node was dependant upon...
+            for field_type in new_leaf.field_types.iter() {
+                let pre_cql_type = map_string_to_cql_type(field_type);
```

Comment:
Ouch, we are parsing the type here? Can't `topo_sort_udts` obtain a list of already parsed UDTs so that parsing multiple times is not necessary?